File: test-submission/test-cf-submit.js
/**
 * Minimal Codeforces Submission Test
 *
 * Uses curl as HTTP transport because Codeforces blocks Node.js
 * HTTP clients (axios, node-fetch) via TLS fingerprinting.
 *
 * Tests the full flow:
 *   1. Validate session by checking logged-in status
 *   2. Fetch CSRF token from submission page
 *   3. Submit a simple solution to a problem
 *   4. Poll for verdict via CF API
 *
 * Usage (full cookie string — recommended):
 *   node test-cf-submit.js --cookies "JSESSIONID=ABC123; cf_clearance=XYZ789; ..." [contestId] [problemIndex]
 *
 * Usage (JSESSIONID only — may be blocked by Cloudflare):
 *   node test-cf-submit.js "ABC123" [contestId] [problemIndex]
 *
 * How to get cookies:
 *   1. Log into codeforces.com in your browser
 *   2. Press F12 → Network tab → refresh page
 *   3. Click the first request → Headers → Cookie
 *   4. Copy the ENTIRE cookie string
 *
 * Example:
 *   node test-cf-submit.js --cookies "JSESSIONID=638EB4DE...; cf_clearance=abc..." 4 A
 */

const { execFileSync } = require('child_process');
const cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');
const os = require('os');

// ─── CONFIG ──────────────────────────────────────────────
let COOKIES = '';
let CONTEST_ID = '4';
let PROBLEM_INDEX = 'A';

// Parse arguments: --cookie-file, --cookies "full string", or just JSESSIONID
if (process.argv[2] === '--cookie-file') {
  // Read cookies from a file (avoids shell escaping issues)
  const cookieFile = process.argv[3] || '';
  COOKIES = fs.readFileSync(cookieFile, 'utf-8').trim();
  CONTEST_ID = process.argv[4] || '4';
  PROBLEM_INDEX = process.argv[5] || 'A';
} else if (process.argv[2] === '--cookies') {
  COOKIES = process.argv[3] || '';
  CONTEST_ID = process.argv[4] || '4';
  PROBLEM_INDEX = process.argv[5] || 'A';
} else if (process.argv[2]) {
  // Legacy: just JSESSIONID value
  COOKIES = `JSESSIONID=${process.argv[2]}`;
  CONTEST_ID = process.argv[3] || '4';
  PROBLEM_INDEX = process.argv[4] || 'A';
}

// Correct solution for 4A (Watermelon)
const SOURCE_CODE = [
  '#include <bits/stdc++.h>',
  'using namespace std;',
  'int main() {',
  '    int n;',
  '    cin >> n;',
  '    if (n > 2 && n % 2 == 0)',
  '        cout << "YES" << endl;',
  '    else',
  '        cout << "NO" << endl;',
  '    return 0;',
  '}',
].join('\n');

const LANGUAGE_ID = '54'; // GNU C++17

// ─── CURL WRAPPER ────────────────────────────────────────

/**
 * Safe curl GET using execFileSync (no shell injection).
 */
function curlGet(url, cookies) {
  const args = ['-s', '-L', '--max-time', '20'];
  if (cookies) {
    args.push('-H', `Cookie: ${cookies}`);
  }
  // Add browser-like headers to avoid Cloudflare blocks
  args.push('-H', 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36');
  args.push('-H', 'Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8');
  args.push('-H', 'Accept-Language: en-US,en;q=0.9');
  args.push('--compressed');
  args.push(url);
  return execFileSync('curl', args, { maxBuffer: 10 * 1024 * 1024 }).toString();
}

/**
 * Safe curl POST using a temp file for the source code body.
 * Returns the response body (follows redirects).
 */
function curlPost(url, formFields, cookies, referer) {
  const args = ['-s', '-L', '--max-time', '30'];
  args.push('-H', `Cookie: ${cookies}`);
  args.push('-H', 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36');
  args.push('--compressed');
  if (referer) {
    args.push('-H', `Referer: ${referer}`);
  }

  // For fields that may contain special characters (like source code),
  // write to a temp file and use @file syntax
  const tmpDir = os.tmpdir();

  for (const [key, val] of Object.entries(formFields)) {
    if (key === 'source') {
      // Write source code to a temp file to avoid shell issues
      const tmpFile = path.join(tmpDir, `cf_source_${Date.now()}.txt`);
      fs.writeFileSync(tmpFile, val);
      args.push('-F', `source=<${tmpFile}`);
      // Clean up after curl finishes (we'll do it after exec)
      args._tmpFile = tmpFile;
    } else {
      args.push('-F', `${key}=${val}`);
    }
  }

  args.push(url);

  try {
    const result = execFileSync('curl', args, { maxBuffer: 10 * 1024 * 1024 }).toString();
    // Clean up temp file
    if (args._tmpFile) {
      try {
        fs.unlinkSync(args._tmpFile);
      } catch (e) {
        /* ignore */
      }
    }
    return result;
  } catch (err) {
    if (args._tmpFile) {
      try {
        fs.unlinkSync(args._tmpFile);
      } catch (e) {
        /* ignore */
      }
    }
    throw err;
  }
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── STEP 1: Validate JSESSIONID ────────────────────────

function validateSession(cookies) {
  console.log('\n[Step 1] Validating session...');
  // Show which cookies we have
  const cookieNames = cookies
    .split(';')
    .map((c) => c.trim().split('=')[0])
    .filter(Boolean);
  console.log(`  Cookies present: ${cookieNames.join(', ')}`);

  const html = curlGet('https://codeforces.com/', cookies);
  const $ = cheerio.load(html);

  // Look for the handle in the header nav
  const handleLink = $('a[href^="/profile/"]').first();
  const handle = handleLink.text().trim();

  if (handle) {
    console.log(`  ✅ Logged in as: ${handle}`);
    return handle;
  }

  // Alternative: check for logout link presence
  if (html.includes('/logout')) {
    const match = html.match(/handle\s*=\s*"([^"]+)"/);
    const h = match ? match[1] : 'unknown';
    console.log(`  ✅ Session valid (handle: ${h})`);
    return h;
  }

  throw new Error('JSESSIONID is invalid or expired. Not logged in.');
}

// ─── STEP 2: Fetch CSRF Token ───────────────────────────

function fetchCsrfToken(cookies, contestId) {
  console.log('\n[Step 2] Fetching CSRF token...');

  const url = `https://codeforces.com/contest/${contestId}/submit`;
  const html = curlGet(url, cookies);
  const $ = cheerio.load(html);

  let csrfToken = $('input[name="csrf_token"]').val();

  if (!csrfToken) {
    csrfToken = $('meta[name="X-Csrf-Token"]').attr('content');
  }

  if (!csrfToken) {
    const match =
      html.match(/name="csrf_token"[^>]*value="([^"]+)"/) ||
      html.match(/value="([^"]+)"[^>]*name="csrf_token"/) ||
      html.match(/csrf_token'\s*,\s*'([a-f0-9]+)'/i);
    csrfToken = match ? match[1] : null;
  }

  if (!csrfToken) {
    console.log('  Page title:', $('title').text());
    console.log('  HTML length:', html.length);
    $('input[type="hidden"]').each((i, el) => {
      console.log(`  Hidden: name=${$(el).attr('name')} val=${$(el).val()?.substring(0, 20)}`);
    });
    throw new Error('Could not extract CSRF token.');
  }

  console.log(`  ✅ CSRF token: ${csrfToken.substring(0, 20)}...`);
  return csrfToken;
}

// ─── STEP 3: Submit Code ────────────────────────────────

function submitCode(cookies, contestId, problemIndex, csrfToken, sourceCode, languageId) {
  console.log(`\n[Step 3] Submitting code to ${contestId}${problemIndex}...`);

  const url = `https://codeforces.com/contest/${contestId}/submit?csrf_token=${csrfToken}`;
  const referer = `https://codeforces.com/contest/${contestId}/submit`;

  const formFields = {
    csrf_token: csrfToken,
    action: 'submitSolutionFormSubmitted',
    ftaa: '',
    bfaa: '',
    contestId: contestId,
    submittedProblemIndex: problemIndex,
    programTypeId: languageId,
    source: sourceCode,
    tabSize: '4',
    sourceFile: '',
  };

  const responseBody = curlPost(url, formFields, cookies, referer);
  const $ = cheerio.load(responseBody);
  const pageTitle = $('title').text().trim();

  // Check for errors
  const errorSpan = $('span.error').text().trim();
  if (errorSpan) {
    console.log(`  ⚠️  Error: ${errorSpan}`);
  }

  // Success indicators
  if (
    pageTitle.includes('Submission') ||
    pageTitle.includes('My') ||
    responseBody.includes('status-cell') ||
    responseBody.includes('view-source') ||
    responseBody.includes('submissionId')
  ) {
    console.log(`  ✅ Submission accepted! (page: "${pageTitle}")`);
    return true;
  }

  if (errorSpan) {
    throw new Error(`Submission rejected: ${errorSpan}`);
  }

  console.log(`  ⚠️  Page: "${pageTitle}" — will verify via API...`);
  return true;
}

// ─── STEP 4: Poll for Verdict ───────────────────────────

async function pollVerdict(handle, contestId, maxAttempts = 20) {
  console.log(`\n[Step 4] Polling for verdict (handle: ${handle})...`);

  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
    try {
      const apiUrl = `https://codeforces.com/api/user.status?handle=${handle}&from=1&count=5`;
      const raw = curlGet(apiUrl);
      const data = JSON.parse(raw);

      if (data.status !== 'OK') {
        console.log(`  Attempt ${attempt}: API status=${data.status}`);
        await sleep(3000);
        continue;
      }

      const ourSub = data.result.find((s) => String(s.contestId) === String(contestId));

      if (!ourSub) {
        console.log(`  Attempt ${attempt}: No submission found for contest ${contestId}`);
        await sleep(3000);
        continue;
      }

      const verdict = ourSub.verdict || 'TESTING';

      if (verdict === 'TESTING' || !ourSub.verdict) {
        console.log(`  Attempt ${attempt}: Testing... (${ourSub.passedTestCount || 0} tests passed)`);
        await sleep(3000);
        continue;
      }

      // Final verdict
      console.log('');
      console.log('  ════════════════════════════════════');
      console.log(`  ║ Submission ID: ${ourSub.id}`);
      console.log(`  ║ Verdict:       ${verdict}`);
      console.log(`  ║ Tests Passed:  ${ourSub.passedTestCount || 0}`);
      console.log(`  ║ Time:          ${ourSub.timeConsumedMillis || 0} ms`);
      console.log(`  ║ Memory:        ${((ourSub.memoryConsumedBytes || 0) / 1024).toFixed(0)} KB`);
      console.log('  ════════════════════════════════════');
      console.log('');

      if (verdict === 'OK') {
        console.log('  🎉 ACCEPTED! Submission flow works correctly.');
      } else {
        console.log(`  ℹ️  Verdict: ${verdict} — but the submission flow itself works!`);
      }

      return ourSub;
    } catch (err) {
      console.log(`  Attempt ${attempt}: Error — ${err.message}`);
      await sleep(3000);
    }
  }

  console.log('  ⚠️  Max attempts reached. Check Codeforces manually.');
  return null;
}

// ─── MAIN ───────────────────────────────────────────────

async function main() {
  if (!COOKIES) {
    console.log(`
╔══════════════════════════════════════════════════════════════╗
║  Codeforces Submission Test                                  ║
╠══════════════════════════════════════════════════════════════╣
║                                                              ║
║  Usage (recommended — full cookies):                         ║
║    node test-cf-submit.js --cookies "COOKIE_STRING" [cid] [p]║
║                                                              ║
║  Usage (simple — may hit Cloudflare):                        ║
║    node test-cf-submit.js <JSESSIONID> [contestId] [idx]     ║
║                                                              ║
║  How to get the full cookie string:                          ║
║    1. Log into codeforces.com                                ║
║    2. Press F12 → Network tab → refresh page                 ║
║    3. Click the first request (codeforces.com)               ║
║    4. In Request Headers, copy the entire Cookie value       ║
║                                                              ║
║  Example:                                                    ║
║    node test-cf-submit.js --cookies \\                        ║
║      "JSESSIONID=ABC; cf_clearance=XYZ" 4 A                  ║
║                                                              ║
╚══════════════════════════════════════════════════════════════╝
`);
    process.exit(1);
  }

  console.log('═══════════════════════════════════════════');
  console.log(' Codeforces Submission Flow Test');
  console.log(`  Contest: ${CONTEST_ID}, Problem: ${PROBLEM_INDEX}`);
  console.log('═══════════════════════════════════════════');

  try {
    const handle = validateSession(COOKIES);
    const csrfToken = fetchCsrfToken(COOKIES, CONTEST_ID);
    submitCode(COOKIES, CONTEST_ID, PROBLEM_INDEX, csrfToken, SOURCE_CODE, LANGUAGE_ID);
    await pollVerdict(handle, CONTEST_ID);
    console.log('\n✅ TEST COMPLETE — The submission pipeline works!');
  } catch (err) {
    console.error(`\n❌ TEST FAILED: ${err.message}`);
    process.exit(1);
  }
}

main();

File: test-submission/test_cf_submit.py
"""
Codeforces Submission Test using Selenium

Uses a real Chrome browser to bypass Cloudflare protection.

Usage:
    python test_cf_submit.py <JSESSIONID> [contest_id] [problem_index]

Example:
    python test_cf_submit.py 638EB4DEF6FC958BACEA3389C85B34C7 4 A
"""

import sys
import time
import json
import requests
import undetected_chromedriver as uc
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait, Select
from selenium.webdriver.support import expected_conditions as EC

# ─── CONFIG ──────────────────────────────────────────────
JSESSIONID = sys.argv[1] if len(sys.argv) > 1 else None
CONTEST_ID = sys.argv[2] if len(sys.argv) > 2 else "4"
PROBLEM_INDEX = sys.argv[3] if len(sys.argv) > 3 else "A"

# Correct solution for 4A (Watermelon)
SOURCE_CODE = r"""
#include <bits/stdc++.h>
using namespace std;
int main() {
    int n;
    cin >> n;
    if (n > 2 && n % 2 == 0)
        cout << "YES" << endl;
    else
        cout << "NO" << endl;
    return 0;
}
""".strip()

LANGUAGE_TEXT = "GNU C++17"  # Text shown in the language dropdown


def create_driver():
    """Create an undetected Chrome driver to bypass Cloudflare."""
    options = uc.ChromeOptions()
    options.add_argument("--headless=new")
    options.add_argument("--no-sandbox")
    options.add_argument("--disable-dev-shm-usage")
    options.add_argument("--disable-gpu")
    options.add_argument("--window-size=1920,1080")

    driver = uc.Chrome(options=options, version_main=145)
    driver.set_page_load_timeout(30)
    return driver


def inject_cookie(driver, jsessionid):
    """Navigate to Codeforces and inject the JSESSIONID cookie."""
    print("\n[Step 0] Setting up browser session...")

    # First navigate to codeforces.com to set the domain
    driver.get("https://codeforces.com/")
    time.sleep(2)

    # Delete any existing JSESSIONID and add ours
    driver.delete_cookie("JSESSIONID")
    driver.add_cookie(
        {
            "name": "JSESSIONID",
            "value": jsessionid,
            "domain": ".codeforces.com",
            "path": "/",
            "secure": True,
        }
    )

    print(f"  ✅ JSESSIONID cookie injected")


def validate_session(driver):
    """Step 1: Check if we're logged in."""
    print("\n[Step 1] Validating session...")

    driver.get("https://codeforces.com/")
    time.sleep(3)

    page_source = driver.page_source

    # Check for Cloudflare challenge
    if "Attention Required" in driver.title or "Just a moment" in driver.title:
        print("  ⏳ Cloudflare challenge detected, waiting...")
        time.sleep(8)
        driver.get("https://codeforces.com/")
        time.sleep(3)

    # Look for the logged-in handle
    try:
        handle_el = driver.find_element(By.CSS_SELECTOR, 'a[href^="/profile/"]')
        handle = handle_el.text.strip()
        if handle:
            print(f"  ✅ Logged in as: {handle}")
            return handle
    except Exception:
        pass

    # Try another way
    if "/logout" in page_source:
        print("  ✅ Session is valid (logged in)")
        # Try to extract handle from page
        try:
            els = driver.find_elements(By.CSS_SELECTOR, ".lang-chooser a")
            for el in els:
                href = el.get_attribute("href") or ""
                if "/profile/" in href:
                    handle = el.text.strip()
                    print(f"  Handle: {handle}")
                    return handle
        except Exception:
            pass
        return "unknown"

    print(f"  Page title: {driver.title}")
    raise Exception("JSESSIONID is invalid or expired. Not logged in.")


def fetch_csrf_and_submit(driver, contest_id, problem_index, source_code):
    """Steps 2 & 3: Get CSRF token and submit code.

    Strategy: Try multiple URLs that have submit forms.
    The problem page has an inline submit form at the bottom.
    If that fails, try the problemset submit page.
    """

    # The pages to try — problem page often has looser Cloudflare rules
    urls_to_try = [
        (
            f"https://codeforces.com/contest/{contest_id}/problem/{problem_index}",
            "problem page",
        ),
        (
            f"https://codeforces.com/problemset/problem/{contest_id}/{problem_index}",
            "problemset page",
        ),
        (f"https://codeforces.com/contest/{contest_id}/submit", "submit page"),
        (
            f"https://codeforces.com/contest/{contest_id}/submit/{problem_index}",
            "submit/problem page",
        ),
    ]

    csrf_token = None
    loaded_url = None

    # ── Step 2: Find a page with a CSRF token ──
    print(f"\n[Step 2] Finding CSRF token...")

    for url, label in urls_to_try:
        print(f"  Trying {label}: {url}")
        driver.get(url)
        time.sleep(4)

        title = driver.title
        if "Attention Required" in title or "Just a moment" in title:
            print(f"    ⚠️ Cloudflare blocked ({title}), trying next...")
            continue

        # Extract CSRF token from the page
        try:
            csrf_input = driver.find_element(
                By.CSS_SELECTOR, 'input[name="csrf_token"]'
            )
            csrf_token = csrf_input.get_attribute("value")
        except Exception:
            # Try meta tag
            try:
                csrf_meta = driver.find_element(
                    By.CSS_SELECTOR, 'meta[name="X-Csrf-Token"]'
                )
                csrf_token = csrf_meta.get_attribute("content")
            except Exception:
                pass

        if not csrf_token:
            # Try extracting from page source via JS
            try:
                csrf_token = driver.execute_script(
                    """
                    var el = document.querySelector('input[name="csrf_token"]');
                    if (el) return el.value;
                    var meta = document.querySelector('meta[name="X-Csrf-Token"]');
                    if (meta) return meta.getAttribute('content');
                    // Try from Codeforces JS variable
                    var match = document.body.innerHTML.match(/csrf='([a-f0-9]+)'/);
                    if (match) return match[1];
                    return null;
                """
                )
            except Exception:
                pass

        if csrf_token:
            print(f"  ✅ CSRF token found: {csrf_token[:20]}...")
            loaded_url = url
            break
        else:
            print(f"    No CSRF token on this page (title: {title})")

    if not csrf_token:
        raise Exception("Could not find CSRF token on any accessible page.")

    # ── Step 3: Submit the code ──
    print(f"\n[Step 3] Submitting code to {contest_id}{problem_index}...")

    # Check if there's a submit form on the current page
    has_submit_form = False
    try:
        driver.find_element(By.NAME, "submittedProblemIndex")
        has_submit_form = True
    except Exception:
        pass

    if has_submit_form:
        # Fill the form on the current page
        return _submit_via_form(driver, contest_id, problem_index, source_code)
    else:
        # Use JavaScript fetch() from the browser context to POST the submission
        # This inherits the browser's cookies and passes Cloudflare
        return _submit_via_js_fetch(
            driver, contest_id, problem_index, source_code, csrf_token
        )


def _submit_via_form(driver, contest_id, problem_index, source_code):
    """Submit using the HTML form on the page."""
    print("  Using HTML form submission...")

    # Select problem
    try:
        problem_select = Select(driver.find_element(By.NAME, "submittedProblemIndex"))
        problem_select.select_by_value(problem_index)
        print(f"  Selected problem: {problem_index}")
    except Exception as e:
        print(f"  ⚠️ Could not select problem (may be pre-selected): {e}")

    # Select language
    try:
        lang_select = Select(driver.find_element(By.NAME, "programTypeId"))
        for option in lang_select.options:
            if "C++17" in option.text and "GNU" in option.text:
                lang_select.select_by_value(option.get_attribute("value"))
                print(f"  Selected language: {option.text}")
                break
        else:
            lang_select.select_by_value("54")
            print("  Selected language: GNU C++17 (by value)")
    except Exception as e:
        print(f"  ⚠️ Could not select language: {e}")

    # Enter source code
    try:
        source_textarea = driver.find_element(By.ID, "sourceCodeTextarea")
        source_textarea.clear()
        source_textarea.send_keys(source_code)
        print("  Entered code in textarea")
    except Exception:
        try:
            toggle = driver.find_element(By.ID, "toggleEditorCheckbox")
            if not toggle.is_selected():
                toggle.click()
                time.sleep(1)
            source_textarea = driver.find_element(By.ID, "sourceCodeTextarea")
            source_textarea.clear()
            source_textarea.send_keys(source_code)
            print("  Entered code in textarea (after toggle)")
        except Exception:
            driver.execute_script(
                """
                var sourceInput = document.querySelector('#sourceCodeTextarea') 
                    || document.querySelector('textarea[name="source"]');
                if (sourceInput) {
                    sourceInput.value = arguments[0];
                    sourceInput.dispatchEvent(new Event('input'));
                } else if (window.ace) {
                    var editors = document.querySelectorAll('.ace_editor');
                    if (editors.length > 0) {
                        ace.edit(editors[0]).setValue(arguments[0]);
                    }
                }
            """,
                source_code,
            )
            print("  Entered code via JavaScript")

    time.sleep(1)

    # Click submit
    try:
        submit_btn = driver.find_element(By.CSS_SELECTOR, 'input[type="submit"].submit')
        submit_btn.click()
    except Exception:
        try:
            submit_btn = driver.find_element(By.CSS_SELECTOR, "#singlePageSubmitButton")
            submit_btn.click()
        except Exception:
            submit_btn = driver.find_element(By.CSS_SELECTOR, 'input[type="submit"]')
            submit_btn.click()

    print("  ✅ Submit button clicked!")
    time.sleep(5)
    return _check_submission_result(driver)


def _submit_via_js_fetch(driver, contest_id, problem_index, source_code, csrf_token):
    """Submit using JavaScript fetch() from within the browser context.
    This inherits all cookies and passes Cloudflare since it comes from a real browser page.
    """
    print("  Using JavaScript fetch() submission (inherits browser session)...")

    submit_url = (
        f"https://codeforces.com/contest/{contest_id}/submit?csrf_token={csrf_token}"
    )

    result = driver.execute_script(
        """
        var url = arguments[0];
        var csrfToken = arguments[1];
        var contestId = arguments[2];
        var problemIndex = arguments[3];
        var source = arguments[4];
        var languageId = arguments[5];

        var formData = new URLSearchParams();
        formData.append('csrf_token', csrfToken);
        formData.append('action', 'submitSolutionFormSubmitted');
        formData.append('ftaa', '');
        formData.append('bfaa', '');
        formData.append('contestId', contestId);
        formData.append('submittedProblemIndex', problemIndex);
        formData.append('programTypeId', languageId);
        formData.append('source', source);
        formData.append('tabSize', '4');
        formData.append('sourceFile', '');

        var xhr = new XMLHttpRequest();
        xhr.open('POST', url, false);  // synchronous
        xhr.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded');
        xhr.setRequestHeader('Referer', 'https://codeforces.com/contest/' + contestId + '/submit');

        try {
            xhr.send(formData.toString());
            return {
                status: xhr.status,
                url: xhr.responseURL,
                body: xhr.responseText.substring(0, 2000),
                ok: true
            };
        } catch(e) {
            return { ok: false, error: e.toString() };
        }
    """,
        submit_url,
        csrf_token,
        contest_id,
        problem_index,
        source_code,
        "54",
    )

    if not result or not result.get("ok"):
        raise Exception(f"XHR submission failed: {result}")

    status = result.get("status")
    resp_url = result.get("url", "")
    body = result.get("body", "")

    print(f"  Response status: {status}")
    print(f"  Response URL: {resp_url}")

    if status == 200 and (
        "my" in resp_url or "status" in resp_url or "submissionId" in body
    ):
        print(f"  ✅ Submission accepted!")
        return True

    if "You have submitted exactly the same code before" in body:
        print("  ✅ Submission mechanism works! (duplicate code detected)")
        return True

    # Check for errors in HTML
    if "error" in body.lower():
        # Try to extract error text
        import re

        err_match = re.search(r'<span class="error[^"]*">([^<]+)</span>', body)
        if err_match:
            print(f"  ⚠️ Error: {err_match.group(1)}")

    print(f"  Body preview: {body[:300]}")
    return True


def _check_submission_result(driver):
    """Check if submission was successful after page load."""
    page_title = driver.title
    page_source = driver.page_source

    errors = driver.find_elements(By.CSS_SELECTOR, "span.error")
    for err in errors:
        err_text = err.text.strip()
        if err_text:
            print(f"  ⚠️ Error: {err_text}")

    if (
        "My submissions" in page_source
        or "status" in driver.current_url
        or "my" in driver.current_url
    ):
        print(f'  ✅ Submission accepted! (page: "{page_title}")')
        return True

    if "You have submitted exactly the same code before" in page_source:
        print("  ✅ Submission mechanism works! (duplicate code detected)")
        return True

    print(f"  Current URL: {driver.current_url}")
    print(f"  Page title: {page_title}")
    return True


def poll_verdict(handle, contest_id, max_attempts=20):
    """Step 4: Poll CF API for submission verdict."""
    print(f"\n[Step 4] Polling for verdict (handle: {handle})...")

    for attempt in range(1, max_attempts + 1):
        try:
            url = (
                f"https://codeforces.com/api/user.status?handle={handle}&from=1&count=5"
            )
            resp = requests.get(url, timeout=10)
            data = resp.json()

            if data.get("status") != "OK":
                print(f"  Attempt {attempt}: API status={data.get('status')}")
                time.sleep(3)
                continue

            # Find our submission for this contest
            our_sub = None
            for sub in data["result"]:
                if str(sub.get("contestId")) == str(contest_id):
                    our_sub = sub
                    break

            if not our_sub:
                print(
                    f"  Attempt {attempt}: No submission found for contest {contest_id}"
                )
                time.sleep(3)
                continue

            verdict = our_sub.get("verdict", "TESTING")

            if verdict == "TESTING" or verdict is None:
                passed = our_sub.get("passedTestCount", 0)
                print(f"  Attempt {attempt}: Testing... ({passed} tests passed)")
                time.sleep(3)
                continue

            # Final verdict
            sub_id = our_sub.get("id")
            tests = our_sub.get("passedTestCount", 0)
            time_ms = our_sub.get("timeConsumedMillis", 0)
            mem_bytes = our_sub.get("memoryConsumedBytes", 0)

            print("")
            print("  ════════════════════════════════════")
            print(f"  ║ Submission ID: {sub_id}")
            print(f"  ║ Verdict:       {verdict}")
            print(f"  ║ Tests Passed:  {tests}")
            print(f"  ║ Time:          {time_ms} ms")
            print(f"  ║ Memory:        {mem_bytes // 1024} KB")
            print("  ════════════════════════════════════")
            print("")

            if verdict == "OK":
                print("  🎉 ACCEPTED! Submission flow works correctly.")
            else:
                print(
                    f"  ℹ️  Verdict: {verdict} — but the submission flow itself works!"
                )

            return our_sub

        except Exception as e:
            print(f"  Attempt {attempt}: Error — {e}")
            time.sleep(3)

    print("  ⚠️ Max attempts reached. Check Codeforces manually.")
    return None


def main():
    if not JSESSIONID:
        print(
            """
╔══════════════════════════════════════════════════════════════╗
║  Codeforces Submission Test (Selenium)                       ║
╠══════════════════════════════════════════════════════════════╣
║                                                              ║
║  Usage:                                                      ║
║    python test_cf_submit.py <JSESSIONID> [contestId] [idx]   ║
║                                                              ║
║  How to get JSESSIONID:                                      ║
║    1. Log into codeforces.com                                ║
║    2. Press F12 → Application → Cookies                      ║
║    3. Find "JSESSIONID" and copy its value                   ║
║                                                              ║
║  Example:                                                    ║
║    python test_cf_submit.py 638EB4DEF6FC958BACEA3389C85B34C7 ║
║                                                              ║
╚══════════════════════════════════════════════════════════════╝
"""
        )
        sys.exit(1)

    print("═══════════════════════════════════════════")
    print(" Codeforces Submission Flow Test (Selenium)")
    print(f"  Contest: {CONTEST_ID}, Problem: {PROBLEM_INDEX}")
    print("═══════════════════════════════════════════")

    driver = None
    try:
        driver = create_driver()

        # Step 0: Inject cookie
        inject_cookie(driver, JSESSIONID)

        # Step 1: Validate session
        handle = validate_session(driver)

        # Steps 2 & 3: Fetch form and submit
        fetch_csrf_and_submit(driver, CONTEST_ID, PROBLEM_INDEX, SOURCE_CODE)

        # Step 4: Poll for verdict
        if handle and handle != "unknown":
            poll_verdict(handle, CONTEST_ID)
        else:
            print("\n[Step 4] Skipping — could not determine handle.")
            print("  Check your submissions on Codeforces manually.")

        print("\n✅ TEST COMPLETE — The submission pipeline works!")

    except Exception as e:
        print(f"\n❌ TEST FAILED: {e}")
        if driver:
            # Save screenshot for debugging
            try:
                driver.save_screenshot("/tmp/cf_debug_screenshot.png")
                print("  Debug screenshot saved to /tmp/cf_debug_screenshot.png")
            except Exception:
                pass
        sys.exit(1)
    finally:
        if driver:
            driver.quit()


if __name__ == "__main__":
    main()
